### PR TITLE
Add primitive floats (binary64 floating-point numbers)

### LINKED ISSUE #8276: Implementing primitive floats (binary64 floating-point numbers)
Dear Coq developers,

We open this issue to keep you informed of some on-going work on implementing primitive floats in Coq.

This work was done by Guillaume Bertholon (@Sakarah) during his internship with Pierre Roux (@proux01) and myself (@erikmd).

The current development can be found here: https://github.com/validsdp/coq/tree/primitive-floats
(it compiles with `./configure -native-compiler no`  (`native_compute` support not yet implemented))

We briefly summarize below the motivations of this work and the design questions we already started tackling:

### Motivations

We'd like to increase Coq computing power to be able to handle complex algorithms in validated numerics (and use them in proofs by reflection). We already see a few developments that could benefit from such primitive floats:
- http://coq-interval.gforge.inria.fr/
- https://github.com/validsdp/validsdp
and this would more generally open the way to Coq-verified versions of algorithms such as the ones implemented in:
- http://www.ti3.tu-harburg.de/rump/intlab/

Currently, the above Coq developments emulate floating-point computations by relying on integer computations. Such operations appear to be about three order of magnitudes slower than primitive FPU operations.

### Workflow

I briefly discussed with @maximedenes at the Coq IW, and he later advised me to base our work upon his primitive-integers implementation (https://github.com/maximedenes/coq/tree/primitive-integers)

Here are the main steps we followed:
1. define a minimal working interface for the IEEE 754-2008 binary64 format
1. define a fully-specified spec w.r.t. a minimal excerpt of Flocq (http://flocq.gforge.inria.fr/)
1. prepare a compatibility layer that could later be added to Flocq
1. hence the choice of the specification type presented below
1. implementation for `compute` and `vm_compute`, at the OCaml and C levels
1. run some benchmarks.

### Focus on the specification type for floats

We discussed with Guillaume Melquiond (@silene) the design of this type and tested several versions.

The current version (file EmulatedFloat.v) is a trade-off. To sum up, it is the same as the `full_float` type in Flocq (file IEEE754/Binary.v) except that NaN payloads are ignored:

```
Inductive emulated_float :=
  | E754_zero : bool -> emulated_float
  | E754_infinity : bool -> emulated_float
  | E754_nan : emulated_float
  | E754_finite : bool -> positive -> Z -> emulated_float.
```

Specification of the operations is then mostly borrowed from Flocq.

Subtle points identified:
- NaNs: their payload is hardware-dependent, meaning some hardware could produce one value and another a different one, which could easily lead to a proof of False
- comparison: do not use IEEE-754 comparison for Leibniz equality (doesn't distinguish +0 and -0 whereas 1/+0 = +oo and 1/-0 = -oo are easily distinguished)
- primitive int63 are unsigned, this does require some care when using them as interface.

@silene also suggested two different versions:

(1)
```
Inductive foo_float : Set :=
  | NaN | PInf | NInf | PZero | NZero
  | Finite (s : bool) (m : positive) (e : Z) : foo_float.
```

and
(2)
```
Inductive foo_float : Set :=
  | NaN | PInf | NInf | PZero | NZero
  | Finite (s : bool) (mh ml e : int31) : foo_float
```

which could be adapted to
(2a)
```
Inductive foo_float : Set :=
  | NaN | PInf | NInf | PZero | NZero
  | Finite (s : bool) (m e : int63) : foo_float
```

but we did not chose these at the current stage of the implementation.
(2) would make the addition more painful to specify with int31/63 (due to wild shiftings).
(1) could be chosen later to avoid boxing/allocating for values +/- Inf and Zero.

Although these may offer performance gains in conversions, this type is expected to be used for specification rather than effective computations.
Maybe another interface should be designed for effective computations (but do we really need it? We already have efficient comparison, is_{nan,zero,infinity} as well as mantissa/exponent extraction and conversion functions).

### Primitive operations

The following primitive operations are implemented (file FloatNative.v):

- `opp : float -> float`
- `abs : float -> float`
- `compare : float -> float -> option comparison`
- `mul, add, sub, div : float -> float -> float`
- `sqrt : float -> float`
- `of_int63 : int -> float`  (rounds value if necessary)
- `normfr_mantissa : float -> int`
  (if input has absolute value in [0.5, 1), return its mantissa, sign is ignored)
- `frshiftexp : float -> float * int`
  (`frshiftexp f` gives `(m, e)` with |m| in [0.5, 1) and f = m * 2^(e-shift))
- `ldshiftexp : float -> int -> float`  (ldshiftexp f e = f * 2^(e-shift))

Anything missing?

The last four operations enable to implement `Prim2EF : float -> emulated_float`
and `EF2Prim : emulated_float -> float`(file FloatOps.v),
used to write the axioms to specify primitive operations (file FloatAxioms.v).

### Benchmarks

We ran some benchmark on a matrix Cholesky decomposition
(file posdef_check.v in https://github.com/validsdp/validsdp/tree/posdef_check)

We roughly observed a speedup of a factor about 50 compared to operations emulated with primitive integers in Coq master. Note that @maximedenes's primitive-integers already offered a speedup of about a factor 2 compared to master.

| matrix    | prim-int | prim-float | speedup |
| --------- | --------:| ----------:| -------:|
| 25 × 25   | 0.051 s  | 0.007 s    | 7.29    |
| 50 × 50   | 0.304 s  | 0.020 s    | 15.20   |
| 75 × 75   | 1.394 s  | 0.051 s    | 27.33   |
| 100 × 100 | 2.738 s  | 0.102 s    | 26.84   |
| 125 × 125 | 4.525 s  | 0.184 s    | 24.59   |
| 150 × 150 | 7.590 s  | 0.299 s    | 25.38   |
| 175 × 175 | 12.008 s | 0.474 s    | 25.33   |
| 200 × 200 | 18.658 s | 0.709 s    | 26.32   |
| 225 × 225 | 25.488 s | 0.967 s    | 26.36   |
| 250 × 250 | 35.593 s | 1.381 s    | 25.77   |
| 275 × 275 | 47.634 s | 1.778 s    | 26.79   |
| 300 × 300 | 62.790 s | 2.345 s    | 26.78   |
| 325 × 325 | 79.595 s | 2.931 s    | 27.16   |
| 350 × 350 | 99.556 s | 3.620 s    | 27.50   |

### Other question

Pretty-printing and parsing of floating-point values raises interesting questions. An easy solution would be to parse/print floats in hexadecimal format (e.g., `0xcp-3`) but this is ugly and perfectly unreadable for humans. A decimal format (e.g., `1.5`) would be more readable.
It is known that using 17 digits while printing binary64 guarantees `parse \circ print : float -> float` to be the identity although `parse` is of course not injective.

Currently `rawnnumeral` interpreters parse integers represented as `string * sign` and interpreters not handling signs just raise an error.
Would it make sense to extend this to decimal numbers as `string * sign * int` (e.g., `1.5` would give `("15", true, -1)` (BTW, the convention for signs here is the reverse of that of the IEEE-754 standard)) such that interpreters only handling integers would just raise an error in case of negative exponent?

### PR DESCRIPTION
**Pull request by erikmd**
**Kind:** feature.

Closes #8276

~~Depends on: PR #8764~~ (merged)

~~Depends on: PR #9905 and PR #9925~~ (merged)

This PR implements primitive floats as described in issue #8276. This work was primarily done by Guillaume Bertholon (@Sakarah) during his internship with Pierre Roux (@proux01) and me (@erikmd). 

### Tasks

- [x] Added / updated test-suite
- [x] Coqchk fixed
- [x] 32 bit implem fixed
- [x] Dune build updated
- [x] ~add convert/round to integer primitives (c.f. https://github.com/coq/coq/issues/8276#issuecomment-414556220)~ [postponed]
- [x] add `=?`, `<?`, `<=?` for primitive floats
- [x] add a module ExtrOCamlFloats (take inspiration from https://github.com/coq/coq/pull/10430)
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in `CHANGES.md`.

### Overlays

- https://github.com/lukaszcz/coqhammer/pull/25
- https://github.com/LPCIC/coq-elpi/pull/58
- https://github.com/unicoq/unicoq/pull/28

### CODE REVIEW

File: kernel/byterun/coq_instruct.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
+/* DO NOT EDIT: automatically generated by kernel/genOpcodeFiles.ml */#pragma once
```
Comment (SkySkimmer) [inline kernel/byterun/coq_instruct.h]: This file shouldn't be committed.

Comment (erikmd) [reply kernel/byterun/coq_instruct.h]: @SkySkimmer Indeed, fixed; thanks for the remark.

**Comment by erikmd** [comment]
The first two commits come from other PRs (resp. #9925 and #9905) and add 113 lines.

Among the remaining 3400 lines added:

- OCaml (and C)
  - add float in kernel 1038
  - vm_compute 274
  - native_compute 336
  - parsing/pretty-printing 190
  - checker 50
- Coq spec 524 (mostly borrowed from Flocq)
- Coq proofs 331
- tests 652
- doc 14 (still to complete)

**Comment by erikmd** [comment]
There is one remaining item to discuss: the interface for comparisons.

Currently, the theory `Float.v` exposes a function

```coq
 Inductive float_comparison : Set :=
     FEq : float_comparison
   | FLt : float_comparison
   | FGt : float_comparison
   | FNotComparable : float_comparison

Parameter compare : float -> float -> float_comparison.
```

The standard way (as per IEEE 754) to implement the equality would be

```coq
Definition eqb f1 f2 := match compare f1 f2 with
  | FEq => true
  | _ => false
  end.
```

but this function (and a corresponding notation `==`) is currently not provided.

Hence the question:
1. Should we keep the current signature of `Float` as is?
2. or add the above eqb (and related functions `<=, <, >=, >`)?
3. or add them as primitive operators?

@maximedenes @vbgl what was your "philosophy" for the related design choice in the Int63 theory?

File: doc/sphinx/language/coq-library.rst [hunk 2]
```diff
@@ -990,7 +991,106 @@ Notation    Interpretation  Precedence  Associativity
...
+
+  Print spec_float.
+  Check Prim2SF.
+  Check mul_spec.
+
+For more details on the available definitions and lemmas, see the
+online documentation of the ``Floats`` library at
+https://coq.inria.fr/distrib/current/stdlib/Coq.Floats.Floats.html.
+
+  .. _userscontributions:
```
Comment (Zimmi48) [inline doc/sphinx/language/coq-library.rst]: It seems that you accidentally added spacing on this line.

**Review by Zimmi48 — COMMENTED** [review]
I am under the impression that the documentation (including the changelog entry) is missing an important part. It makes it seem like this is only an addition to the standard library, when it is also an addition to the kernel. It is also inconsistent with the primitive integer documentation which is a section in the “Extensions of Gallina” chapter. I suggest moving the changelog entry in the kernel section, adding a (small) Primitive floats section after the Primitive Integers one (or changing the section to be about Primitive integers *and* floats), and to add cross-references with the section in the library chapter.

**Comment by erikmd** [comment]
@Zimmi48 thanks for your suggestions! I've updated the doc accordingly (then rebased on latest master).

File: doc/changelog/01-kernel/09867-floats.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+- Add primitive floats implementing binary64 floating-point arithmetic,
+  see :ref:`primitive-floats`
```
Comment (Zimmi48) [inline doc/changelog/01-kernel/09867-floats.rst]: IMHO this is an achievement that deserves a bit more showing-off, i.e. a tag longer changelog entry.

File: doc/sphinx/language/coq-library.rst [hunk 4]
```diff
@@ -990,6 +991,106 @@ Notation    Interpretation  Precedence  Associativity
...
+``_ * _``    ``mul``
+``_ / _``    ``div``
+``_ == _``   ``eqb``
+``_ < _``    ``ltb``
+``_ <= _``   ``leb``
+``_ ?= _``   ``compare``
+===========  ==============
+
+Floating-point constants are parsed and pretty-printed as (17-digit)
+decimal constants. This ensures that the composition `parse \o print`
```
Comment (Zimmi48) [inline doc/sphinx/language/coq-library.rst]: What is this `\o`? Did you mean to use a LaTeX operator? If yes, you could use `:math:`. You can also use unicode.

File: doc/sphinx/language/coq-library.rst [hunk 5]
```diff
@@ -990,6 +991,106 @@ Notation    Interpretation  Precedence  Associativity
...
+
+.. coqtop:: all
+
+  Print spec_float.
+  Check Prim2SF.
+  Check mul_spec.
+
+For more details on the available definitions and lemmas, see the
+online documentation of the ``Floats`` library at
+https://coq.inria.fr/distrib/current/stdlib/Coq.Floats.Floats.html.
```
Comment (Zimmi48) [inline doc/sphinx/language/coq-library.rst]: I am against giving such precise links that point to the "current" stdlib doc when we should actually point to the version of the doc for the same version of Coq. So until #10142 is fixed, I would actually remove the link.

File: doc/sphinx/language/gallina-extensions.rst [hunk 6]
```diff
@@ -2449,6 +2449,48 @@ extraction. Instead, it has to be provided by the user (if they want to compile
...
+   Primitive mul := #float64_mul.
+   Notation "x * y" := (mul x y) : float_scope.
+
+   Axiom mul_spec : forall x y, Prim2SF (x * y)%float = SF64mul (Prim2SF x) (Prim2SF y).
+
+where :g:`Prim2SF` is defined in the :g:`FloatOps` module.
+
+The set of such operators is described in section :ref:`floats_library`.
+
+These primitive declarations are regular axioms. As such, they must be trusted and are listed by the
```
Comment (Zimmi48) [inline doc/sphinx/language/gallina-extensions.rst]:
```suggestion
These primitive declarations are regular axioms. As such, they must be trusted, and are listed by the
```

**Review by Zimmi48 — APPROVED** [review]

**Comment by maximedenes** [comment]
@proux01 do you think we can target 8.11 for this? It would mean that we get this ready before Nov 1st. I can review it, but I've noticed there are two boxes in the PR description that are not checked yet.

**Comment by proux01** [comment]
I hope so. As for the two remaining checkboxes:
* extraction should be relatively easy and we'll definitely get it by Nov 1st (in the meantime it should not be an obstacle to a review)
* round to integer is a relatively minor feature and at worst we can drop it for 8.11

File: test-suite/primitive/float/frexp.v [hunk 7]
```diff
@@ -0,0 +1,30 @@
+(* TODO add tests for ldexp (particularly with overflow with 31 and 63 bits integers) *)
```
Comment (silene) [inline test-suite/primitive/float/frexp.v]: I guess you meant `frexp` rather than `ldexp`. Even so, I do not understand the comment.

**Review by silene — CHANGES_REQUESTED** [review]
My only concern is with `configure.ml`. Other than that, everything looks good to me.

File: test-suite/primitive/float/frexp.v (hunk 7 above)
Comment (proux01) [reply test-suite/primitive/float/frexp.v]: If I remember correctly, we experienced a bug with `ldexp` giving infinities instead of zero or vice versa because of overflows when manipulating the exponent before calling `ldshiftexp`. The bug was fixed but I should had test cases for that.

Comment (erikmd) [reply test-suite/primitive/float/frexp.v]: Hi @silene, thanks for your remark! we just looked at `ldexp` with @proux01 for adding tests, which uncovered a buggy behavior, which is now fixed hopefully.
All changes are gathered in the last commit https://github.com/coq/coq/pull/9867/commits/f53b44e65e458b230d248be3ec0b0a8981095752 .
(We rebased just to fix the conflict and relaunch the CI.)

File: configure.ml [hunk 8]
```diff
@@ -456,7 +456,7 @@ let coq_bin_annot_flag = if !prefs.bin_annot then "-bin-annot" else ""
 let coq_safe_string = "-safe-string"
 let coq_strict_sequence = "-strict-sequence"
 
-let cflags = "-Wall -Wno-unused -g -O2"
+let cflags = "-Wall -Wno-unused -g -O2 -fexcess-precision=standard -msse2 -mfpmath=sse"
```
[the remark this answers is not in the document]
Comment (proux01) [reply configure.ml]: > What are Coq's supported architectures?

We could not find any indication in the documentation (README.md, INSTALL, sphinx doc).
Maybe is it time to add something. I don't know what would be the best place for this indication though. @Zimmi48 any idea?

> make sure that the SSE2 flags are only passed if the compiler is able to recognize them.

Done. The main incompatibility we are aware of is x87 which is the default for both OCaml and gcc on x86_32, leading to double roundings. The logic we just implemented in `configure.ml` is the following:
* test for SSE2 support to add corresponding flags to `CFLAGS`
* test for double rounding in OCaml
  * if no double rounding happens, fine
  * if double rounding happens and SSE2 is available, fine (it will be used, preventing double rounding)
  * otherwise, abort configure

In case miscompilation still happens, a runtime test (at the end of `kernel/float64.ml` could still catch it).

The code is in the following commit: https://github.com/coq/coq/pull/9867/commits/158561390aa496b42c2bf993629e39de77bdbe62 (we had to rebase for CI)

Currently, the test for SSE2 availability is done by checking the presence of `#define __SSE2_MATH__ 1` in the output of `cc -march=native -dM -E -msse2 -mfpmath=sse some_file.c` @MSoegtropIMC isn't it to Unix specific? what about Windows?

[the remark this answers is not in the document]
Comment (SkySkimmer) [reply configure.ml]: If we need to mess with c flags it also needs to be done in the dune file btw

[the remark this answers is not in the document]
Comment (Zimmi48) [reply configure.ml]: > We could not find any indication in the documentation (README.md, INSTALL, sphinx doc).
> Maybe is it time to add something. I don't know what would be the best place for this indication though. @Zimmi48 any idea?

Sure it would seem like something sensible to document. I suggest doing it in another PR, and indeed `INSTALL` seems like the reasonable place for this. Note though how some stuff like the VM and native compute machinery have fallback mechanisms that make the use of Coq possible even when these features cannot be used.

[the remark this answers is not in the document]
Comment (maximedenes) [reply configure.ml]: > these features

Which features do you mean?

[the remark this answers is not in the document]
Comment (Zimmi48) [reply configure.ml]: VM and native compute.

File: kernel/float64.ml [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+(*** Test at runtime that no harmful double rounding seems to
+   be performed with an intermediate 80 bits representation (x87). *)
+let () =
+  let b = ldexp 1. 53 in
+  let s = add 1. (ldexp 1. (-52)) in
+  if add b s <= b || add b 1. <> b then
+    failwith "Detected double-rounding due to the use of intermediate \
+              80-bit floating-point representation. Use of Float is \
+              thus unsafe."
```
Comment (silene) [inline kernel/float64.ml]: Just to be on the safe side, I would also test that gradual underflow behaves properly, in case someone runs Coq on a crippled processor (or on SSE2 with flush-to-zero enabled for "performance" reasons).

File: configure.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (proux01) [reply configure.ml]: > Sure it would seem like something sensible to document. I suggest doing it in another PR, and indeed `INSTALL` seems like the reasonable place for this.

I honestly don't know what to require in the general case. Thus, I just add the requirement for IEEE-754 arch in INSTALL in the current PR.

File: kernel/float64.ml (hunk 9 above)
Comment (proux01) [reply kernel/float64.ml]: Very good point, done (https://github.com/coq/coq/pull/9867/commits/7555656c423dd41dfc91ed5f8ac71c0e7638cf94)

File: INSTALL [hunk 10]
```diff
@@ -21,6 +21,10 @@ WHAT DO YOU NEED ?
 
    - a C compiler
 
+   - an IEEE-754 compliant architecture with rounding to nearest
+     ties to even as default rounding mode (most architectures
+     should work nowadays)
```
Comment (Zimmi48) [inline INSTALL]: What happens if the architecture is not compliant?

Comment (proux01) [reply INSTALL]: At best the configure phase detects it and aborts. At worst, one may be able to prove False on this architecture. Do you think this is worth mentioning (maybe in a paragraph below)?

Comment (Zimmi48) [reply INSTALL]: Yes, absolutely worth mentioning!

File: test-suite/primitive/float/frexp.v (hunk 7 above)
Comment (erikmd) [reply test-suite/primitive/float/frexp.v]: A proof update was missing in the commit "Fix ldshiftexp". I've just force-pushed to amend that commit accordingly (whose SHA1 is now dba99578b488bbc6efb89623965ebda3cef97eb9).

File: INSTALL (hunk 10 above)
Comment (erikmd) [reply INSTALL]: Hi @Zimmi48 @proux01,
so I've just pushed a commit with one such mention (in the `INSTALL` file):
fbeb6eefb74c18230366d2ef6643b49f72ed21b3

**Review by Zimmi48 — APPROVED** [review]
Additions to `INSTALL` look good to.

**Comment by erikmd** [comment]
Hi,  the [previous CI build](https://gitlab.com/coq/coq/-/jobs/329190452) failed but the issue didn't show up when I built the branch, so I had to rebase on master to reproduce and fix it.
To sum up, `kernel/float64.ml` was unnecessarily depending on `lib/util.ml` (just to use `pervasives_compare`) and this triggered an issue when building `votour`. Now, `kernel/float64.ml` just contains [that line](https://github.com/coq/coq/blob/8d9647538edf0240d6ee96e7284340d4cd32d52c/kernel/float64.ml#L54-L55).

Meanwhile, I verified that each commit still builds and just found two small issues (now fixed):

1. some `Require Import FloatClass.` was committed too early in the commit history.

2. some flags were missing for the `votour` build rule in `Makefile.build`, triggering some linking issue.  
    (FYI the URL of the corresponding patch is: https://github.com/coq/coq/pull/9867/commits/379dfab0c6efebc4efd58cb9215159516d557316#diff-8162d704df7537f6cf8008829973520f )

Hopefully the CI will be green now :)

@maximedenes and @silene, could one of you take a look at the last commit that has been added since your review? (namely, extraction for primitive floats in 8d9647538edf0240d6ee96e7284340d4cd32d52c).

**Comment by silene** [comment]
Seems fine to me, but since extraction depends on a file internal to Coq, should there not be some documentation (e.g., directly in the `.v` file) on how to recover it and compile/link it?

**Comment by erikmd** [comment]
Dear @silene, good point; I've added a coqdoc comment in `ExtrOcamlFloats.v` (feel free to proofread); and I tweaked again the `Makefile.build` file so the Azure build failure on macOS should hopefully be fixed now…

Guillaume, would you agree to be the assignee of this PR?

**Comment by silene** [comment]
Sure. As far as I can tell, the only remaining piece is to understand what the dune story is.

**Comment by ejgallego** [comment]
> @ejgallego : how would you advise to communicate the `CFLAGS` elaborated in `configure.ml` to dune?

The usual method is to generate a file with the flags, then include it.

File: dune [hunk 11]
```diff
@@ -1,10 +1,13 @@
 ; Default flags for all Coq libraries.
 (env
- (dev     (flags :standard -rectypes -w -9-27+40+60 \ -short-paths))
+ (dev     (flags :standard -rectypes -w -9-27+40+60 \ -short-paths)
+          (c_flags (:include dune.c_flags)))
```
[the remark this answers is not in the document]
Comment (proux01) [reply dune]: Indeed, done

There seems to be many places in the CI where dune is run without having run configure first. I'm trying to fix that.

**Comment by SkySkimmer** [comment]
How important is this flag manipulation? If someone compiles without it do they get inconsistencies?

**Comment by silene** [comment]
The flag is critical for x86-32 compilers. But the user won't experience any inconsistency (hopefully). Indeed, the commit also includes some runtime checks, which means that Coq will terminate abruptly if the discrepancy is detected. In other words, there will be no inconsistencies because Coq will be unusable.

**Comment by erikmd** [comment]
Dear @silene and @proux01, FYI:

* I added yet another commit to better document the extraction (w.r.t. float literals),
* and I rebased on latest master to fix the conflict (Pierre had added a commit to fix the compilation of `RMicromega.v` but this has been fixed independently in #11021, so that commit became unneeded).

File: dune (hunk 11 above)
[the remark this answers is not in the document]
Comment (proux01) [reply dune]: I gave up. There was a hack in `config/dune`, I just copy/pasted it in `kernel/byterun/dune` which seems to do the job for now.

**Comment by proux01** [comment]
> the only remaining piece is to understand what the dune story is.

@silene : this is now fixed, so looks like we are all set (the CI failure for fcsl-pcm seems unrelated)

**Review by maximedenes — APPROVED** [review]
LGTM, modulo CI. If successful, I'll try to merge today.

**Comment by maximedenes** [comment]
> I'll try to merge today

(if @silene is not available, I meant)

**Comment by proux01** [comment]
CI is ok (fcsl-pcm is unrelated and windows64 and fiat-crypto are system errors (no space on disk))

**Comment by ppedrot** [comment]
Not wanting to be pushy, but this needs to get in before tonight or be postponed to 8.12.

**Review by SkySkimmer — DISMISSED** [review]
IIUC this makes it necessary to invoke configure before dune
Is that OK @ejgallego ?
If yes build-system.dune.md must be updated.

**Comment by proux01** [comment]
> IIUC this makes it necessary to invoke configure before dune

No, that's fixed, I copied/pasted a hack from @ejgallego so that dune correctly calls configure itself. That's as neat as it was previously and CI is happy.

To sum up, all points have been addressed and CI is OK so looks like this can be merged.

**Comment by maximedenes** [comment]
> fcsl-pcm is unrelated

@ppedrot is this failure known?

**Comment by SkySkimmer** [comment]
It is known, see recent gitter, tldr they need to adapt to a recent 
mathcomp change

Gaëtan Gilbert

**Comment by maximedenes** [comment]
Ok, thanks! Merging, then.

**Comment by proux01** [comment]
Thanks!

**Comment by maximedenes** [comment]
Thanks for this nice piece of work!

**Comment by proux01** [comment]
Congrats to @Sakarah who made a large part of it!
And thanks again to everybody who reviewed or commented.

**Comment by herbelin** [comment]
Sorry to bother for a naive question: I realize that under MacOS X, I get the message "clang: warning: argument unused during compilation: '-fexcess-precision=standard'" when compiling the C files in `kernel/byterun`. Do I have to worry or is it expected?

**Comment by proux01** [comment]
This option prevents use of 80 bit registers on old 32 bits x86 architectures, so it's indeed unused most of the time.

There is a test at configure time and, as a last resort, at runtime to detect the use of those registers. So nothing to worry about in terms of soundness.

**Comment by herbelin** [comment]
Thanks for the reply. Is there then something specific to improve with MacOS X configuration?

**Comment by proux01** [comment]
If we can live with the warning, I would say no.

We might be able to silence it, but that would require extra code and imply extra risks of miscompilation. Not worth it IMHO.

**Comment by ejgallego** [comment]
What risks would fixing the warning imply? Shouldn't just be a matter of using the right set of flags for each compiler?

**Comment by proux01** [comment]
> What risks would fixing the warning imply?

The risk of not using a required flag, possibly making Coq unusable. That sound much harmful than using a useless flag.

**Comment by ejgallego** [comment]
> The risk of not using a required flag, possibly making Coq unusable. That sound much more harmful than using a useless flag.

I don't see why the flag would not be used in compilers that support it.

**Comment by proux01** [comment]
Because the architecture makes it meaningless (at least, that's what clang warned about).

**Comment by ejgallego** [comment]
> Because the architecture makes it meaningless (at least, that's what clang warned about).

The warning seems about an unrecognized flag [docs support this https://clang.llvm.org/docs/ClangCommandLineReference.html]

IMO using a compiler flag that doesn't exist is a serious bug in our configure.
